float32 x_offset
float32 y_offset